tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +227,280 @@ static bool HasCanonicalDimensionNumbers(
+
+// Looks through a set of dimension that has been marked as Reduction axes
+// if it is found within the set, then we set it as "reduction", otherwise
+// we can label it as "parallel"
+SmallVector<StringRef, 3> GetEinsumLoopsAttrs(
+    std::set<std::string> input_ind, std::set<std::string> reduction_dims) {
+  SmallVector<StringRef, 3> res;
+  for (std::string dim : input_ind) {
+    if (reduction_dims.find(dim) == reduction_dims.end())
+      res.push_back(getParallelIteratorTypeName());
+    else
+      res.push_back(getReductionIteratorTypeName());
+  }
+  return res;
+}
+
+// Check for where each dimension of the output operand came from
+// If that dimension is dynamically sized then we push it to dyn_size vector
+SmallVector<Value, 2> ExtractDynamicEinsumSizes(
+    OpBuilder& b, Location loc, Value lhs, Value rhs,
+    std::vector<std::string> lhs_loop_vec,
+    std::vector<std::string> rhs_loop_vec,
+    std::vector<std::string> output_loop_vec) {
+  std::vector<Value> tensor_vec = {lhs, rhs};
+  std::vector<ArrayRef<int64_t>> tensor_shape_vec;
+  for (auto tensor_val : tensor_vec) {
+    auto tensorCur_type = tensor_val.getType().dyn_cast<RankedTensorType>();
+    if (!tensorCur_type) return {};
+    tensor_shape_vec.push_back(tensorCur_type.getShape());
+  }
+  SmallVector<Value, 2> dyn_sizes;
+  for (std::string const& dim_ind : output_loop_vec) {
+    int arg_number = 0;
+    auto dim_ind_it =
+        std::find(lhs_loop_vec.begin(), lhs_loop_vec.end(), dim_ind);
+    auto dim_ind_pos = dim_ind_it - lhs_loop_vec.begin();
+    if (dim_ind_it == lhs_loop_vec.end()) {
+      arg_number = 1;
+      dim_ind_it = std::find(rhs_loop_vec.begin(), rhs_loop_vec.end(), dim_ind);
+      dim_ind_pos = dim_ind_it - rhs_loop_vec.begin();
+    }
+    if (tensor_shape_vec[arg_number][dim_ind_pos] != ShapedType::kDynamicSize)
+      continue;
+    dyn_sizes.push_back(
+        b.create<memref::DimOp>(loc, tensor_vec[arg_number], dim_ind_pos));
+  }
+  return dyn_sizes;
+}
+
+// Adds indices/axes that are missing from output set
+std::set<std::string> FindSummationAxes(std::set<std::string> input_set,
+                                        std::set<std::string> output_set) {
+  std::set<std::string> summation_axes;
+  for (auto ind : input_set) {
+    if (output_set.find(ind) == output_set.end()) summation_axes.insert(ind);
+  }
+  return summation_axes;
+}
+
+// Given a 1:1 map from std::string -> affine dimension expression
+// we can get the affine expression of dimensions that an
+// operand will access based on the input_str of einsum_config
+// For example:
+// let string_dim_umap = {'a' : d0, 'b' : d1, 'c' : d2}
+// for einsum_config "abc,cb->acb"
+// the first_input_operand will get umap[{"a","b","c"}] -> (d0, d1, d2)
+// the second_input_operand will get umap[{"c","b"}] -> (d2, d1)
+// the ouput_operand will get umap[{"a","c","b"}] -> (d0, d2, d1)
+SmallVector<AffineExpr, 4> getExprFromConfig(
```

Comment:
This can be `SmallVector<AffineExpr>`, I think it's 4 by default.

Reply:
style nit: `GetExprFromConfig`

Reply:
done